### PR TITLE
[ios] Base implementation of iCloud Synchronization

### PR DESCRIPTION
**Pull request by kirylkaveryn**
This PR contains the basic implementation of iCloud support for iOS.
This feature will allow user to have up-to-date bookmarks on all Apple devices by using iCloud Documents storage.

Things that are still in progress:
- unit tests for the state resolving 
- error handling
- merge conflicts and corner cases

### CODE REVIEW

**Comment by biodranik** [comment]
If I created an empty list on one device, I would expect it to be present on another device too. What are the issues with syncing empty lists?

**Comment by biodranik** [comment]
Am I correctly understanding that the current code always processes all files, even if user edited one file? That could be a huge battery issue in some cases when users have tens/hundreds of bookmarks. Note that we have many existing and potential users with tons of bookmarks that are collected from 2012, when this feature was released.

**Comment by kirylkaveryn** [comment]
> Am I correctly understanding that the current code always processes all files, even if user edited one file? That could be a huge battery issue in some cases when users have tens/hundreds of bookmarks. Note that we have many existing and potential users with tons of bookmarks that are collected from 2012, when this feature was released.

Code iterates through file's urls lightweight attributes (metadata). And check is there are some `modifiedDate` updates. If there are - only this file will be updated/saved.
It needs to test the battery consumption. But it does not change the core code at all.

**Comment by kirylkaveryn** [comment]
> If I created an empty list on one device, I would expect it to be present on another device too. What are the issues with syncing empty lists?

My `SynchronizarionStateManager` is fully isolated from the core code (except the isLocalIsEmpty hardcoded line). So if the array of elements should be empty for this case.
I want to replace hardcoded `localContentIsEmpty` with `localContent.isEmpty` (because it is actually empty) to isolate the state from the BookmarksManager at all.
So I can move it upper to the LocalDirectoryMonitor, but I still cannot skip this hardcoded line...
```swift
     // this place makes my code dependent from the BookmarksManager:
    let localContentIsEmpty = BookmarksManager.shared().sortedUserCategories()
.first(where: { BookmarksManager.shared().category(withId: $0.categoryId).bookmarksCount != 0}) == nil 

    switch (localContentIsEmpty /* <--- */ , cloudContents.isEmpty) {
    case (true, true):
      outgoingEvents = []
    case (true, false):
      outgoingEvents = cloudContents.notTrashed.map { .createLocalItem($0) }
    case (false, true):
      outgoingEvents = localContents.map { .createCloudItem($0) }
    case (false, false):
      if isInitialSynchronization {
        outgoingEvents.append(contentsOf: resolveInitialSynchronizationConflicts(localContents: localContents, cloudContents: cloudContents))
      }
```

**Comment by kirylkaveryn** [comment]
Rebased on master...

**Comment by kirylkaveryn** [comment]
Some features added:
1. New setting in the Settings screen. When there is no bookmarks (empty app) there will be no alert. If user already have bookmarks the alert with shown:
image
And when user taps "OK" share extensions will be triggered to save bookmark. 
Only after this events user will be allowed to enable iCloud.

Settings should be enabled on every device separately to give user ability to save bookmarks on all devices.

2. New method to the BookmarksManager were added: 
`IsEmpty` and `shareAllCatoriesWithCompletion`.
The second one is created to skip usage of observer pattern (because it seems like overhead for such operation when you just call one method and will receive asynchronous results).

Note:
Transactions for the alert is in progress...

**Comment by matheusgomesms** [comment]
Probably closes #2678

**Comment by biodranik** [comment]
@kirylkaveryn please rebase and check if anything else can be merged separately. Any additional review and testing are welcomed.

@dvdmrtnz @fabwu @euf @vng @rtsisyk

File: iphone/Maps/PrivacyInfo.xcprivacy [hunk 1]
```diff
@@ -23,6 +23,7 @@
 		<dict>
 			<key>NSPrivacyAccessedAPITypeReasons</key>
 			<array>
+				<string>C617.1</string>
```
Comment (biodranik) [inline iphone/Maps/PrivacyInfo.xcprivacy:26]: This should go to `AccessedAPICategoryFileTimestamp`, right?

Comment (kirylkaveryn) [reply iphone/Maps/PrivacyInfo.xcprivacy:26]: Thanks! This code is set incorrectly here. I'll fix it. This is for the timestamp

File: iphone/Maps/Core/iCloud/DefaultLocalDirectoryMonitor.swift [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+private extension FileManager {
+  func source(for directory: URL) throws -> DispatchSourceFileSystemObject {
+    if !fileExists(atPath: directory.path) {
+      do {
+        try createDirectory(at: directory, withIntermediateDirectories: true)
+      } catch {
+        throw error
+      }
+    }
+    let directoryFileDescriptor = open(directory.path, O_EVTONLY)
```
Comment (biodranik) [inline iphone/Maps/Core/iCloud/DefaultLocalDirectoryMonitor.swift:204]: This code monitor only directory changes, not file changes. E.g. if you edit/overwrite some file in-place, it won't fire. If you add a new file, delete, change file name, then it fires.

Our bookmarks saving code should create a new temporary file, and move it to overwrite the old file atomically, that's why it works.

However, there could be some other cases when files are just overwritten/appended, without modifying the directory.

Another case to test: what if user uploads some good/bad kml files (and other types too) using Finder/File Sharing into OM's monitored folder?

Comment (kirylkaveryn) [reply iphone/Maps/Core/iCloud/DefaultLocalDirectoryMonitor.swift:204]: I wrote the test case to check a file contents update isolated from the bookmark manager.
This code is triggered on _every write operation_ into the directory, not only the creation/renaming/removal.

image



> Another case to test: what if user uploads some good/bad kml files (and other types too) using Finder/File Sharing into OM's monitored folder?

We can only add files to the iCloud local container.

- Good kmls will be parsed and displayed immediately
- Bad kmls will not be displayed because they will not be parsed - I mentioned in todo that we should add an alert to handle this situation and notify a user.
- Other files (not *.kml) will not be synced (until we implement the support of the nested files/directories).

Comment (biodranik) [reply iphone/Maps/Core/iCloud/DefaultLocalDirectoryMonitor.swift:204]: Is any appending write to the existing file also catched? Can a test for append write be added?

Is it possible to avoid detection of appending data?

Mu Files app shows non-kml files. Is it a bug that they are synced to iCloud?

image

Comment (kirylkaveryn) [reply iphone/Maps/Core/iCloud/DefaultLocalDirectoryMonitor.swift:204]: > Is any appending write to the existing file also catched? Can a test for append write be added?

is this not appending writing?
image

> Is it possible to avoid detection of appending data?

Why should we do that?

> My Files app shows non-kml files. Is it a bug that they are synced to iCloud?

It seems that these are temporary kmls created by the core...
I'll add filtering to these files.

**Comment by itfarrier** [comment]
> @itfarrier @fabwu do you want to take a quick look? I know, it is huge. Architecture is good here, but I am very well familiar with all the details of iOS libraries.

@rtsisyk, sorry, but I'm a newbie here. Despite the code I saw demos and I can say that everything's works well. Maybe only production using can give us more information, edge cases, bugs etc.

**Comment by vng** [comment]
Should merge https://github.com/organicmaps/organicmaps/pull/8040 first, right?

**Comment by fabwu** [comment]
I've never worked with the libraries that are used here so I didn't look at the code. However, I tested the sync between my Mac and my iPad and it worked great over Wifi.

@kirylkaveryn Do you know if the sync affects mobile data usage? 

Some people might have many bookmarks so there quite some MBs to sync each time you add a bookmark. Maybe we should couple it to the OM use mobile data setting?

**Comment by kirylkaveryn** [comment]
> Do you know if the sync affects mobile data usage?

There are no straight ways to measure the mobile data usage or the other traffic related to the iCloud syncing. 
The iCloud overall delays and slows down the data usage when the device isn't on wifi. And we cannot force uploading during mobile data usage. We just write the files into the dir that will be synced when the system decides to do that. While testing I didn't face any abnormal data usage. Even files approx 50-60mb are synced really fact.

**Review by fabwu — APPROVED** [review]

**Comment by itfarrier** [comment]
> > > @itfarrier @fabwu do you want to take a quick look? I know, it is huge. Architecture is good here, but I am very well familiar with all the details of iOS libraries.
> > 
> > 
> > @rtsisyk, sorry, but I'm a newbie here. Despite the code I saw demos and I can say that everything's works well. Maybe only production using can give us more information, edge cases, bugs etc.
> 
> Please test on your multiple devices.

Could anyone help me with prod builds for this? I don't have Apple Developer Account for now.

**Comment by fabwu** [comment]
> Could anyone help me with prod builds for this? I don't have Apple Developer Account for now.

Did you try [this](https://github.com/organicmaps/organicmaps/blob/master/docs/INSTALL.md#ios-app)?

I created a dev build for my iPhone and one for my Mac to test this.

**Comment by kirylkaveryn** [comment]
@MontakOleg can you please, if you have free time, to review this huge one? 
There is a lot of code and I believe many things that can be refactored.

**Comment by biodranik** [comment]
Rebase?

**Comment by matheusgomesms** [comment]
Is the "iCloud Synchronization (Beta)" string temporary? Didn't see it in the translations, don't know if it was intentional.

**Comment by kirylkaveryn** [comment]
> Is the "iCloud Synchronization (Beta)" string temporary? Didn't see it in the translations, don't know if it was intentional.

Strings were added and generated (see the last two commits). But we will change the message to more self-describing.

**Comment by matheusgomesms** [comment]
I didn't find the "iCloud Synchronization (Beta)" string, but I suppose it will be replaced to the others already translated in the last commits. Thanks!

**Comment by kirylkaveryn** [comment]
> I didn't find the "iCloud Synchronization (Beta)" string, but I suppose it will be replaced to the others already translated in the last commits. Thanks!

Oh sorry. You are right! This is my mistake(( I forgot to localize this one.

**Comment by fabwu** [comment]
@kirylkaveryn Just that you know: the sync between different iCloud accounts is also possible. I've shared the kml-file with my girlfriend via iCloud and moved it to the Organicmaps folder on her device. We can now share bookmarks with each other and it works great. Thanks for this amazing feature!

**Comment by biodranik** [comment]
@fabwu how exactly did you do that? Selected a file in your OM's bookmarks folder, shared it, then it was added to another account's iCloud, and then you moved the file in another account into OM's bookmarks folder, correct?

**Comment by fabwu** [comment]
> @fabwu how exactly did you do that? Selected a file in your OM's bookmarks folder, shared it, then it was added to another account's iCloud, and then you moved the file in another account into OM's bookmarks folder, correct?

Yes that's how I did it. I've just tested it over wifi so I can't tell how the sync works over the mobile network.
